pallets/corporate-actions/src/distribution.rs

```diff
@@ -0,0 +1,506 @@
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Start and attach a capital distribution, to the CA identified by `ca_id`,
+        /// with `amount` funds in `currency` withdrawn from `portfolio` belonging to `origin`'s DID.
+        ///
+        /// The distribution will commence at `payment_at` and expire at `expires_at`,
+        /// if provided, or if `None`, then there's no expiry.
+        ///
+        /// ## Arguments
+        /// - `origin` which must be a signer for the CAA of `ca_id`.
+        /// - `ca_id` identifies the CA to start a capital distribution for.
+        /// - `portfolio` specifies the portfolio number of the CAA to distribute `amount` from.
+        /// - `currency` to withdraw and distribute from the `portfolio`.
+        /// - `amount` of `currency` to withdraw and distribute.
+        /// - `payment_at` specifies when benefits may first be pushed or claimed.
+        /// - `expires_at` specifies, if provided, when remaining benefits are forfeit
+        ///    and may be reclaimed by `origin`.
+        ///
+        /// # Errors
+        /// - `UnauthorizedAsAgent` if `origin` is not `ticker`'s sole CAA (owner is not necessarily the CAA).
+        /// - `DistributingAsset` if `ca_id.ticker == currency`.
+        /// - `ExpiryBeforePayment` if `expires_at.unwrap() <= payment_at`.
+        /// - `NowAfterPayment` if `payment_at < now`.
+        /// - `NoSuchCA` if `ca_id` does not identify an existing CA.
+        /// - `NoRecordDate` if CA has no record date.
+        /// - `RecordDateAfterStart` if CA's record date > payment_at.
+        /// - `InsufficientPortfolioBalance` if `portfolio` has less than `amount` of `currency`.
+        /// - `InsufficientBalance` if the protocol fee couldn't be charged.
+        #[weight = 2_000_000_000]
+        pub fn distribute(
+            origin,
+            ca_id: CAId,
+            portfolio: PortfolioNumber,
+            currency: Ticker,
+            amount: T::Balance,
+            payment_at: Moment,
+            expires_at: Option<Moment>,
+        ) {
+            // Ensure CA's asset is distinct from the distributed currency.
+            ensure!(ca_id.ticker != currency, Error::<T>::DistributingAsset);
+
+            // Ensure that any expiry date doesn't come before the payment date.
+            ensure!(!expired(expires_at, payment_at), Error::<T>::ExpiryBeforePayment);
+
+            // Ensure `now <= payment_at`.
+            ensure!(<Checkpoint<T>>::now_unix() <= payment_at, Error::<T>::NowAfterPayment);
+
+            // Ensure origin is CAA, `ca_id` exists, and that its a benefit.
+            // and that sufficient funds exist.
+            let caa = <CA<T>>::ensure_ca_agent(origin, ca_id.ticker)?;
+            let from = PortfolioId::user_portfolio(caa, portfolio);
+            let caa = caa.for_event();
+            let ca = <CA<T>>::ensure_ca_exists(ca_id)?;
```

Comment:
You need to check if the CAA's secondary key has "portfolio permission" for this portfolio.